[Docs] Create CC requirements docs (#4637)
Cog-Creators/Red-DiscordBot: docs/guide_cog_creators.rst

----------------------------------
Creating a Cog Creator Application
----------------------------------

.. note::
  You will need to have created and published your cogs before you create a Cog Creator Application!
  See `guide_cog_creation` and `guide_publish_cogs` for more information.

Cog Creator Applications are hosted on the `cogboard <https://cogboard.red/c/apps/12>`__.
To create an application, start a new topic in the "Applications" category and fill out all of the required information.
QA reviews Cog Creator Applications for security and functionality on a first come, first serve basis.
Once your application is reviewed, you will have 14 days to make any requested changes, or to check in with the member of QA who is reviewing your application.

-----------------------------
Requirements for Cog Creators
-----------------------------

The following is a list of the requirements for approved Cog Creators.
QA uses this list to request changes for Cog Creator Applications.
Handling these requirements before submitting your application can streamline the review process.

> Any Cog Creator that does not follow these requirements will have their repo removed from approved listings any may have their Cog Creator status revoked.

Reviewer: ```suggestion
Any Cog Creator that does not follow these requirements will have their repo removed from approved listings and may have their Cog Creator status revoked.
```

- Readme that contains

  - Repository Name
  - Installation Instructions
  - Setup Instructions (if applicable)

>   - Credits (if applicable)

Reviewer: No need to use title casing here. It would also be clearer if "setup instructions" were "extra setup instructions":
```suggestion
  - Repository name
  - Installation instructions
  - Extra setup instructions (if applicable)
  - Credits (if applicable)
```

> - Repo-wide info.json with the keys

Reviewer: ```suggestion
- Repo-wide ``info.json`` file with the keys
```

Author: Edited `files` -> `file` since there is only one repo `info.json` file

  - ``author``
  - ``name``
  - ``short``
  - ``description``

> - Cog info.jsons with the keys

Reviewer: ```suggestion
- Cog ``info.json`` files with the keys
```

  - ``author``
  - ``name``
  - ``short``
  - ``requirements`` (if applicable)
  - ``description``

>   See `info-json-format` for more information on how to set up info.json files.

Reviewer: ```suggestion
  See `info-json-format` for more information on how to set up ``info.json`` files.
```

- No cog contains malicious code.
- No cog contains code that could impact the stability of the bot, such as blocking the event loop for an extended period of time, unreasonably high IO usage, etc.
- No cog contains copied code that does not respect the license of the source.

> - Disclose in the install_msg info.json key of each cog that contains any of the following

Reviewer 2: following -> following:

> - Disclose in the install_msg info.json key of each cog that contains any of the following:

Reviewer: ```suggestion
- Disclose in the ``install_msg`` key of the ``info.json`` file of each cog that contains any of the following:
```

  - Heavy memory or I/O usage.
  - Any NSFW material.
  - Bundled data.
  - Stored data (outside of using Config).
  - Interactions with outside services.
  - Any extra setup instructions required.

- No cog breaks the Discord TOS.

> - No cog conflicts with any core cogs (E.G., causing a core cog to fail to load) unless it is intended to replace that cog.

Reviewer: ```suggestion
- No cog conflicts with any core cogs (e.g., causing a core cog to fail to load) unless it is intended to replace that cog.
```

> - Your repo shows an understanding of a range of common cog practices. This requirement is to ensure QA can trust the safety and functionality of any future code you will create. This is handled on a case-by-case basis, however the following points should outline what we are looking to see:

Reviewer: ```suggestion
- Your repo shows an understanding of a range of common cog practices. This requirement exists to ensure QA can trust the safety and functionality of any future code you will create. This is handled on a case-by-case basis, however the following points should outline what we are looking to see:
```

>   - Cogs that are more than what is able to be run in an eval.

Reviewer: Perhaps this was supposed to be "a simple eval" or something? I can do a lot of things in an eval...
```suggestion
  - Cogs that are more than what is able to be run in a simple eval.
```

  - Cogs that are more than just a simple API access request.
  - Cogs that properly use Red utilities, including Config, checks, and any other utility functions.
  - Cogs that use event listeners (bot.wait_for or cog-wide listeners) or custom tasks that are efficient and handle exceptions appropriately.
  - Cogs that handle errors properly.
  - Cogs that handle permissions properly.

  *While we're looking for these things to qualify your code, you don't need to apply all of them in order to qualify.*

- Any unusable or broken commands or cogs are hidden.
- The default locale must be English.
- The main cog class and every command must have a doc-string.

> - No cog allows for escalation of permissions. (E.G., a sending a mass ping through the bot without having permission to do so)

Reviewer: ```suggestion
- No cog allows for escalation of permissions. (e.g., sending a mass ping through the bot without having permission to do so)
```

- Respect the role hierarchy. Don’t let a lower role have a way to grant a higher role.
- If your cog install comes with any pre-packaged data, use `bundled_data_path()` to access it.
- If your cog install creates any non-config data, use `cog_data_path()` to store it.
- Unless the cog is intentionally designed to listen to certain input from bots, cogs should ignore input from bots.
- Cogs use public methods of Red where possible.
- Use the proper method if one exists. (and ask for one if it doesn't exist)

  - If that's not possible, don't break anything in core or any other cog with your code.

>   - If you have to use private methods, lock the cog to specific Red versions you can guarantee it works on without breaking anything using the ``min_bot_version`` and ``max_bot_version`` info.json keys.

Reviewer: ```suggestion
  - If you have to use private methods, lock the cog to specific Red versions you can guarantee it works on without breaking anything using the ``min_bot_version`` and ``max_bot_version`` keys in that cog's ``info.json`` file.
```

  - Checking that you aren't reacting to a bot message (``not message.author.bot``) early on.

- Use .gitignore (or something else) to keep unwanted files out of your cog repo.
- Put a license on your cog repo.

  - By default, in most jurisdictions, without a license that at least offers the code for use,
    users cannot legally use your code.

- Use botwide features when they apply. Some examples of this:

  - ``ctx.embed_color``
  - ``bot.is_automod_immune``

- Use checks to limit command use when the bot needs special permissions.
- Check against user input before doing things. Common things to check:

  - Resulting output is safe.
  - Values provided make sense. (eg. no negative numbers for payday)
  - Don't unsafely use user input for things like database input.

- Check events against `bot.cog_disabled_in_guild() <RedBase.cog_disabled_in_guild()>`\

  - Not all events need to be checked, only those that interact with a guild.
  - Some discretion may apply, for example,
    a cog which logs command invocation errors could choose to ignore this
    but a cog which takes actions based on messages should not.

> - Respect settings when treating non command messages as commands.

Reviewer 2: non command -> non-command

  - Don't do unexpected things with user data.
  - Don't expose user data to additional audiences without permission.
  - Don't collect data your cogs don't need.
  - | Don't store data in unexpected locations.
    | Utilize the cog data path, Config, or if you need something more
      prompt the owner to provide it.

- Utilize the data deletion and statement APIs

  - See `redbot.core.commands.Cog.red_delete_data_for_user()`
  - | Make a statement about what data your cogs use with the module level
      variable ``__red_end_user_data_statement__``.
    | This should be a string containing a user friendly explanation of what data
      your cog stores and why.

- Set contextual locales in events and other background tasks that use i18n APIs

  - See `redbot.core.i18n.set_contextual_locales_from_guild()`
  - Usage of i18n APIs within commands automatically has proper contextual locales set.

----------------------------
Perks of being a Cog Creator
----------------------------

- Added to a growing, curated list of approved repositories hosted on the `Red Index <https://index.discord.red/>`__.
- The Cog Creator role on the main Red Server and the Cog Support Server.

> - Access to an additional testing channel and the Advanced Coding channel on the main Red Server.

Reviewer: Perhaps we should just use the actual channel name?
```suggestion
- Access to an additional testing channel and the #advanced-coding channel on the main Red Server.
```

> - Write permission in the V3-Burndown channel on the main Red Server.

Reviewer: Perhaps we should just use the actual channel name?
```suggestion
- Write permission in the #v3-burndown channel on the main Red Server.
```
